File: py/day1.py
import re

f = open("./data/day1.txt", "r")

data = f.readlines()

def get_digits(x):
    x_rev = x[::-1]
    d1 = re.findall(r'\d{1}', x)[0]
    d2 = re.findall(r'\d{1}', x_rev)[0] 

    ret = int(d1 + d2)

    return ret

p1 = sum([get_digits(x) for x in data])

# part 2

repl_dict = {'one': '1', 'two': '2', 'three': '3', 'four': '4', 'five': '5', 'six': '6', 'seven': '7', 'eight': '8', 'nine': '9'}

p = 'one|two|three|four|five|six|seven|eight|nine|1|2|3|4|5|6|7|8|9'

def replace_digits(x):
    for k, v in repl_dict.items():
        x = x.replace(k, v)
    return x

def extract_digits(x, pattern):
    p1 = re.findall(pattern, x)[0]
    p2 = re.findall(pattern[::-1], x[::-1])[0][::-1]
    tmp = [p1, p2]
    ds = [replace_digits(j) for j in tmp]
    ret = int(ds[0] + ds[1])
    return ret

p2 = sum([extract_digits(i, p) for i in data])

File: jl/day1.jl
f = open("./data/day1.txt", "r")

data = readlines(f)

function get_digits(x::String, rx::Regex, o::Bool)
    tmp = [m.match for m in eachmatch(rx, x, overlap=o)]
    d = tmp[begin] * tmp[end]
    ret = parse(Int, d)
    return ret
end

rx1 = r"[0-9]"

p1 = get_digits.(data, rx1, false)

sum(p1)

# p2 ---------

repl_dict = Dict(
    "one" => "1",
    "two" => "2",
    "three" => "3",
    "four" => "4",
    "five" => "5",
    "six" => "6",
    "seven" => "7",
    "eight" => "8",
    "nine" => "9"
)

tst = "eightwo"

rx2 = r"one|two|three|four|five|six|seven|eight|nine|[0-9]"

function get_digits2(x::String, rx::Regex, o::Bool)
    tmp = [m.match for m in eachmatch(rx, x, overlap=o)]
    tmp2 = [tmp[i] in keys(repl_dict) ? repl_dict[tmp[i]] : tmp[i] for i in eachindex(tmp)]
    d = tmp2[begin] * tmp2[end]
    ret = parse(Int, d)
    return ret
end

p2 = get_digits2.(data, rx2, true)

sum(p2)